core/include/detray/navigation/direct_navigator.hpp

```diff
@@ -0,0 +1,372 @@
+            assert(detail::is_invalid_value(static_cast<nav_link_type>(v)) ||
+                   v < detector().volumes().size());
+            m_volume_index = static_cast<nav_link_type>(v);
+        }
+
+        DETRAY_HOST_DEVICE
+        inline auto abort() -> bool {
+            m_status = navigation::status::e_abort;
+            m_heartbeat = false;
+            return m_heartbeat;
+        }
+
+        /// @returns current detector volume of the navigation stream
+        DETRAY_HOST_DEVICE
+        inline auto get_volume() const {
+            return tracking_volume<detector_type>{*m_detector, m_volume_index};
+        }
+
+        /// Set direction
+        DETRAY_HOST_DEVICE
+        inline void set_direction(const navigation::direction dir) {
+            m_direction = dir;
+        }
+
+        DETRAY_HOST_DEVICE
+        inline void set_no_trust() { return; }
+
+        DETRAY_HOST_DEVICE
+        inline void set_full_trust() { return; }
+
+        DETRAY_HOST_DEVICE
+        inline void set_high_trust() { return; }
+
+        DETRAY_HOST_DEVICE
+        inline void set_fair_trust() { return; }
+
+        /// Intersection candidate
+        candidate_t m_candidate;
+        candidate_t m_candidate_prev;
+
+        /// Detector pointer
+        const detector_type *m_detector{nullptr};
+
+        /// Index in the detector volume container of current navigation volume
+        nav_link_type m_volume_index{0u};
+
+        /// Target surfaces
+        sequence_t m_sequence;
+
+        // iterator for forward direction
+        typename sequence_t::iterator m_it;
+
+        // iterator for backward direction
+        typename sequence_t::reverse_iterator m_it_rev;
+
+        /// The navigation direction
+        navigation::direction m_direction{navigation::direction::e_forward};
+
+        /// The navigation status
+        navigation::status m_status{navigation::status::e_unknown};
+
+        /// Step size when the valid intersection is not found for the target
+        scalar_type safe_step_size = 10.f * unit<scalar_type>::mm;
+
+        /// Heartbeat of this navigation flow signals navigation is alive
+        bool m_heartbeat{false};
+    };
+
+    template <typename track_t>
+    DETRAY_HOST_DEVICE inline void init(const track_t &track, state &navigation,
+                                        const navigation::config &cfg,
+                                        const context_type &ctx) const {
+
+        if (navigation.is_complete()) {
+            navigation.m_heartbeat = false;
+            return;
+        }
+
+        navigation.m_heartbeat = true;
+        navigation.update_candidate(!navigation.is_init());
+        update(track, navigation, cfg, ctx);
+    }
+
+    template <typename track_t>
+    DETRAY_HOST_DEVICE inline bool update(
+        const track_t &track, state &navigation, const navigation::config &cfg,
+        const context_type &ctx = {},
+        const bool is_before_actor_run = true) const {
```

Comment:
@niermann999 N.B) I added a new boolean argument to `update` function not to call `next` after the `run_actors` in `propagate_step`. I had to add the boolean argument to the normal navigator but it is not used there.